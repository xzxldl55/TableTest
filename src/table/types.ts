/* eslint-disable @typescript-eslint/no-explicit-any */

import type { PropOptions, PropType } from 'vue-types/dist/types'
type Prop<T, D = T> = PropOptions<T, D> | PropType<T>
type PublicRequiredKeys<T> = {
  [K in keyof T]: T[K] extends { required: true } ? K : never
}[keyof T]

type PublicOptionalKeys<T> = Exclude<keyof T, PublicRequiredKeys<T>>
type InferPropType<T> = T extends null
  ? any // null & true would fail to infer
  : T extends { type: null | true }
    ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
    : T extends ObjectConstructor | { type: ObjectConstructor }
      ? Record<string, any>
      : T extends BooleanConstructor | { type: BooleanConstructor }
        ? boolean
        : T extends Prop<infer V, infer D>
          ? unknown extends V
            ? D
            : V
          : T

// eslint-disable-next-line @typescript-eslint/ban-types
export type IxPublicPropTypes<O> = O extends object
  ? { [K in PublicRequiredKeys<O>]: InferPropType<O[K]> } & { [K in PublicOptionalKeys<O>]?: InferPropType<O[K]> }
  : { [K in string]: any }

export interface paginationType {
  pageIndex: number
  pageSize: number
}

export type tableState = {
  data: Record<string, any>[]
  filters: string[]
  columns: ColumnPublicProps[]
  showHeader: boolean
}

export type TSortState = {
  name: string
  direction: 0 | 1
}

export type TFilterState = {
  filter: string
  searchValue: string
}

export interface TPageConf extends paginationType {
  count: number // 数据条目总数
}

// Props 定义在这里
export const tableProps = {
  data: { // 表格数据
    type: Array as PropType<Record<string, any>>,
    default: (): Array<Record<string, any>> => {
      return []
    },
    required: true,
  },
  columns: { // 列数据
    type: Array as PropType<ColumnPublicProps[]>,
    default: (): ColumnPublicProps[] => {
      return []
    },
    required: true,
  },
  showHeader: { // 是否显示表头
    type: Boolean,
    default: true,
  },
  pagination: { // 分页配置
    type: [Object as PropType<paginationType>, false],
    default: (): paginationType | false => {
      return false
    },
  },
  maxHeight: { // 表格体最大高度配置（默认按照内容撑开即高度为auto，配置后超出限制则滚动页面）
    type: [Number, String],
    default: -1,
  },
  filters: { // 可筛选字段
    type: Array,
    default: (): Array<string> => {
      return []
    },
  },
}
export const columnProps = {
  name: { // 列名 --> 即唯一标识id
    type: String,
    default: '',
    required: true,
  },
  title: { // 列标题
    type: String,
    default: '',
  },
  canSort: {
    type: Boolean,
    default: false,
  },
  renderFn: {
    type: Function || undefined,
    default: undefined,
  },
  sortState: {
    type: Object as PropType<TSortState>,
    default: (): TSortState => {
      return { name: '', direction: 0 }
    },
  },
}


export type TablePublicProps = IxPublicPropTypes<typeof tableProps>
export type ColumnPublicProps = IxPublicPropTypes<typeof columnProps>



