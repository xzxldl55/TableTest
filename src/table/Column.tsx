import { defineComponent } from '@vue/composition-api';
import { columnProps } from './types';

export default defineComponent({
    name: 'TableColumn',
    props: columnProps,
    setup (props, { slots, emit }) {
        const title = props.title
        const name = props.name
        const canSort = props.canSort
        const emitSort = () => {
            emit('click', name);
        }
        return () => {
            return (
                <th column-name={name}>
                    { title }
                    {
                        canSort ?
                        <span class="column-sort"
                              onClick={emitSort}></span> :
                        ''
                    }
                </th>
            )
        }
    }
})