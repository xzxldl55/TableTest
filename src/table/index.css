.left {
    text-align: left;
}
.right {
    text-align: right;
}
.center {
    text-align: center;
}
.mr-1 {
    margin-right: 1rem;
}

/* 表格体 */
table {
    table-layout: fixed;
    border: 1px solid #d7dde4;
    font-size: 1rem;
    width: 100%;
    border-radius: 4px;
    box-sizing: border-box;
    color: #657180;
}
table > thead {
    background: #f5f7f9;
    color: #657180;
}
table > tbody {
    display: inline-block;
    width: 100%;
    overflow: auto;
}
tr {
    display: flex;
    width: 100%;
}
th, td {
    display: inline-block;
    flex: 1;
    white-space: nowrap;
    overflow: hidden;
    text-overflow: ellipsis;
    padding: 8px 12px;
    text-align: center;
}

/* 表头 */
.table-search-container {
    line-height: 32px;
    padding: 4px 0;
}
.table-search-container .table-search-select {
    line-height: 32px;
    height: 32px;
    width: 80px;
    border: 1px solid #dcdfe6;
    padding: 0 16px 0 8px;
    box-sizing: border-box;
    color: #606266;
}
.table-search-container .table-search-input {
    height: 32px;
    box-sizing: border-box;
    border: 1px solid #dcdfe6;
    border-left-color: transparent;
    color: #606266;
    outline: none;
    transition: .5s;
}
.table-search-container .table-search-input:hover, .table-search-input:focus {
    border: 1px solid #39f;
}

/* 列头 */
.column-sort {
    position: relative;
    display: inline-block;
    margin-left: 5px;
    cursor: pointer;
}
.column-sort::before { /* 升序图标 */
    position: absolute;
    top: -5px;
    content: '';
    border: 5px solid transparent;
    border-top-color: #657180;
}
.column-sort::after { /* 降序图标 */
    position: absolute;
    top: -18px;
    content: '';
    border: 5px solid transparent;
    border-bottom-color: #657180;
}
.column-sort.asc::before {
    border-top-color: #39f;
}
.column-sort.des::after {
    border-bottom-color: #39f;
}


/* 分页器 */
.pagination {
    padding: 8px 16px;
    border-top: 1px solid #e3e8ee;
}
.pagination-btn {
    display: inline-block;
    width: 32px;
    height: 32px;
    text-align: center;
    line-height: 32px;
    background-color: #fff;
    color: #657180;
    border: 1px solid #d7dde4;
    margin-right: 8px;
    border-radius: 4px;
    cursor: pointer;
    transition: .5s;
}
.pagination-btn:last-child {
    margin-right: 0;
}
.pagination-btn:hover {
    border-color: #39f;
    color: #39f;
}
.pagination-btn.on {
    background: #39f;
    color: #fff;
    border-color: transparent;
}