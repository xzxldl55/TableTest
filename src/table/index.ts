/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './Table'
import TestColumn from './Column'

export {
    TestTable,
    TestColumn
}
