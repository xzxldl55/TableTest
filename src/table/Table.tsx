/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent, Ref, ref, watch } from '@vue/composition-api'
import { TablePublicProps, tableProps, paginationType } from './types'
import Column from './Column'
import { sortBy } from 'lodash-es'
import './index.css'

export default defineComponent({
  name: 'Table',
  props: tableProps,
  setup(props: TablePublicProps, { slots, emit }) {
    const classes = useClasses(props)
    const style = useStyle(props)
    let sortConf = ref({
      sortKey: '',
      sort: false
    }); // 当前表格排序配置（升序/降序）
    let pagination: paginationType | false = props.pagination as unknown as paginationType | false;
    let tableData = useTableData(props, pagination, sortConf) // 获取table数据【关联props.data, 分页，排序】

    // 侦听分页器变化，当外部分页器变化时重新覆盖组件内部分页器
    watch(() => props.pagination, (val) => pagination = (val as unknown as paginationType | false))

    const count = computed(() => { // 数据条目总数
      if (Array.isArray(props.data)) {
        return props.data.length
      }
      return 0
    })
    const size = computed(() => { // 总页数
      if (pagination) {
        return Math.ceil(count.value / pagination.pageSize)
      }
      return 0
    })

    // 排序
    const sortTable = (name: string) => {

      // 回到首页
      if (pagination) {
        pagination.pageIndex = 1
      }

      // 当前sortBy非name，则初始化按照name升序排序
      if (sortConf.value.sortKey !== name) {
        sortConf.value.sortKey = name
        sortConf.value.sort = true
      } else { // 正按照当前列排序，将排序方式取反
        sortConf.value.sort = !sortConf.value.sort
      }

      // 抛出排序事件
      emit('sort', { name, sort: sortConf.value.sort })
    }

    // 切换页码
    const changePage = (index: number) => {
      if (!pagination) {
        return
      }
      if (pagination.pageIndex === index) {
        return
      }
      pagination.pageIndex = index;

      // 页码切换事件
      emit('page-change', index)
    }


    return () => {
      const operators = slots.operator
      const tableColumns = props.columns || []
      return (
        <table class={classes.value}>
          {/* 表头 */}
          <thead vShow={props.showHeader}>
            <tr>
              {
                tableColumns.map(column => {

                  // TODO：如何消灭IxPublicPropTypes带来的undefind类型
                  return slots[column.name] ?
                    <th>{ slots[column.name]({ column }) }</th> :
                    <Column title={column.title}
                            name={column.name}
                            canSort={column.canSort}
                            onClick={sortTable} />
                })
              }
              {
                operators ?
                <th>操作</th> :
                ''
              }
            </tr>
          </thead>
          {/* 表格主体 */}
          <tbody style={style.value}>
            {
              tableData.value.map(item => {
                return (<tr>
                    {
                      tableColumns.map(column => {
                        return <td column-data-name={column.name}>{ column.renderFn ? column.renderFn(item[column.name]) : item[column.name] }</td>
                      })
                    }
                    <td class="center">
                      {/* TODO: 在Linux上运行时报错不给使用 ?. 语法，待确认是什么语法插件没配上 */}
                      { operators ? operators({ item }) : '' }
                    </td>
                  </tr>)
              })
            }
          </tbody>
          {/* 分页器 */}
          <div class="pagination center">
            {
              pagination ? 
              (
                new Array(size.value).fill(1).map((v, i) => {
                  return <button class={['pagination-btn', { 'on': pagination.pageIndex === i + 1 }]}
                                 onClick={changePage.bind(null, i + 1)}>{ i + 1 }</button>
                })
              ) :
              ''
            }
          </div>
        </table>
      )
    }
  },
})

// 类名处理
function useClasses (props: TablePublicProps) {
  return computed(() => {
    return {
      'simple-table': true,
      'test-class': props.test,
    }
  })
}

// tBody样式处理
function useStyle (props: TablePublicProps) {
  return computed(() => {
    return {
      maxHeight: typeof props.maxHeight === 'string' ? props.maxHeight : props.maxHeight + 'px'
    }
  })
}

// 返回应渲染的Table数据（在函数内进行对数据处理，如分页、排序等）
function useTableData (props: TablePublicProps, pagination: paginationType | false, sortConf: Ref<{ sortKey: string; sort: boolean }>) {
  return computed(() => {
    if (!props.data) {
      return []
    }
    if (!pagination) {
      return props.data
    }

    // ... TODO：ect 其他功能添加，如 已选列表prop的处理

    const start = (pagination.pageIndex - 1) * pagination.pageSize
    const end = start + pagination.pageSize

    // 检测是否需要按某列排序
    if (sortConf.value.sortKey) {
      return sortConf.value.sort ? 
        sortBy(props.data, (obj: { [x: string]: any }) => obj[sortConf.value.sortKey]).slice(start, end) :
        sortBy(props.data, (obj: { [x: string]: any }) => obj[sortConf.value.sortKey]).reverse().slice(start, end)
    }
    return props.data.slice(start, end)
  })
}