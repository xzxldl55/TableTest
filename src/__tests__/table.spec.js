import { mount } from '@vue/test-utils'
import { TestTable } from '../table/index'

const staffMockData = [
  {
    name: 'Alice',
    age: 29,
    sex: 0,
  },
  {
    name: 'Tom',
    age: 27,
    sex: 1,
  },
  {
    name: 'Jerry',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry1',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry2',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry3',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry4',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry5',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry6',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry7',
    age: 7,
    sex: 0,
  },
  {
    name: 'Jerry8',
    age: 5,
    sex: 0,
  },
  {
    name: 'Jerry9',
    age: 12,
    sex: 0,
  },
  {
    name: 'Jerry10',
    age: 88,
    sex: 0,
  },
  {
    name: 'Jerry11',
    age: 77,
    sex: 0,
  },
  {
    name: 'Jerry12',
    age: 66,
    sex: 0,
  },
  {
    name: 'Jerry13',
    age: 26,
    sex: 0,
  },
  {
    name: 'Jerry14',
    age: 24,
    sex: 0,
  },
  {
    name: 'Jerry15',
    age: 13,
    sex: 0,
  },
  {
    name: 'Jerry16',
    age: 31,
    sex: 0,
  },
  {
    name: 'Jerry17',
    age: 33,
    sex: 0,
  },
  {
    name: 'Jerry18',
    age: 22,
    sex: 0,
  },
  {
    name: 'Jerry19',
    age: 17,
    sex: 0,
  },
  {
    name: 'Jerry20',
    age: 16,
    sex: 0,
  },
  {
    name: 'Jerry21',
    age: 15,
    sex: 0,
  },
  {
    name: 'Jerry22',
    age: 14,
    sex: 0,
  },
  {
    name: 'Jerry23',
    age: 13,
    sex: 0,
  },
  {
    name: 'Jerry24',
    age: 12,
    sex: 0,
  },
  {
    name: 'Jerry25',
    age: 11,
    sex: 0,
  },
]
const tableColumns = [
  {
    name: 'name',
    title: '姓名',
  },
  {
    name: 'age',
    title: '年龄',
    canSort: true,
  },
  {
    name: 'sex',
    title: '性别',
    renderFn: val => {
      return val ? '男' : '女'
    },
  },
]
const tablePagination = {
  pageSize: 10, // 每页大小
  pageIndex: 1, // 当前页码
}
const defaultProps = {
  data: staffMockData,
  columns: tableColumns,
  pagination: tablePagination,
  maxHeight: 240,
  showHeader: true,
}

describe('Table', () => {
  const TableMount = options => mount(TestTable, options)
  test('开始测试', () => {
    expect(1).toBe(1)
  })

  test('render Default', () => {
    const wrapper = TableMount({ props: defaultProps })
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('props test', async () => {
    const wrapper = TableMount({
      props: {
        test: true,
        ...defaultProps,
      },
    })

    expect(wrapper.find('.test-class').exists()).toBeTruthy()

    // 1. 分页配置测试 --> 校验pageIndex/pageSize设置结果
    expect(wrapper.findAll('.simple-table > tbody > tr').length).toEqual(10)
    expect(wrapper.find('.pagination-btn.on').html()).toEqual('1')

    // 2. maxHeight最大高度测试，获取Tbody实际高度
    expect(wrapper.find('.simple-table > tbody').attributes('style')).toContain('max-height: 240px')
    await wrapper.setProps({ props: { maxHeight: -1 } })
    expect(wrapper.find('.simple-table > tbody').attributes('style')).not.toContain('max-height: 240px')

    // 3. showHeader，尝试find('table > thead')，查看其display是否为none
    expect(wrapper.find('.simple-table > thead').element.getClientRects.length).toEqual(1)
    await wrapper.setProps({ props: { showHeader: false } })
    expect(wrapper.find('.simple-table > thead').element.getClientRects.length).toEqual(0)

    // 4. 列数据测试，传入renderFn，校验实际结果
    expect(wrapper.findAll('.simple-table td[column-data-name=sex]').filter(node => !['男', '女'].includes(node.html())).length).toEqual(0)
  })

  test('sort data', () => {
    const wrapper = TableMount({
      props: defaultProps,
    })

    {
      // 按年龄排序
      wrapper.find('.simple-table th[column-name=age] .column-sort').click() // 模拟点击排序
      const ageData = Array.from(wrapper.findAll('.simple-table td[column-data-name=age]')).map(node => Number(node.innerHTML))
      const ageSortBeforeStr = ageData.toString()

      // 检查是否升序排序了
      expect(ageSortBeforeStr).toEqual(ageData.sort((a, b) => a - b).toString())
    }

    {
      // 按年龄排序
      wrapper.find('.simple-table th[column-name=age] .column-sort').click() // 模拟点击排序
      const ageData = Array.from(wrapper.findAll('.simple-table td[column-data-name=age]')).map(node => Number(node.innerHTML))
      const ageSortBeforeStr = ageData.toString() // 将结果转换成字符串方便比较

      // 检查是否降序排序了
      expect(ageSortBeforeStr).toEqual(ageData.sort((a, b) => b - a).toString())
    }
  })

  test('pagination test', () => {
    const wrapper = TableMount({
      props: defaultProps,
    })

    // 检测默认配置
    expect(wrapper.find('.simple-table .pagination-btn.on').html()).toEqual('1')
    expect(wrapper.findAll('.simple-table tbody > tr').length).toEqual(10)

    // 切换页码，分页大小
    wrapper.setProps({ props: { pagination: { pageIndex: 2, pageSize: 12 } } })
    expect(wrapper.find('.simple-table .pagination-btn.on').html()).toEqual('2')
    expect(wrapper.findAll('.simple-table tbody > tr').length).toEqual(12)
  })
})
