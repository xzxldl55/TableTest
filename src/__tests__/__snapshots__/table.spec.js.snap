// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<table class=\\"\\">
  <thead>
    <tr></tr>
  </thead>
  <tbody style=\\"max-height: -1px;\\"></tbody>
  <div class=\\"pagination center\\"></div>
</table>"
`;
